[Safer-CPP] Update LLVM revision without restarting buildbot

briannafan opened the pull request:
#### e88ca68561d247ac9b9fb93405d794be22089cfe
<pre>
[Safer-CPP] Update LLVM revision without restarting buildbot
<a href="https://bugs.webkit.org/show_bug.cgi?id=298691">https://bugs.webkit.org/show_bug.cgi?id=298691</a>
<a href="https://rdar.apple.com/160330585">rdar://160330585</a>

Reviewed by Aakash Jain.

Add GetLLVMVersion to read the canonical LLVM version from a file.
This allows us to update the file and update which version of clang
we&apos;re using without restarting buildbot.

* Tools/CISupport/Shared/steps.py:
(GetLLVMVersion):
(GetLLVMVersion.__init__):
(GetLLVMVersion.run):
(GetLLVMVersion.getResultSummary):
(CheckOutLLVMProject.run_vc):
(CheckOutLLVMProject.doStepIf):
(UpdateClang.run):
(UpdateClang.doStepIf):
(PrintClangVersion.run):

* Tools/CISupport/Shared/steps_unittest.py:
(TestPrintClangVersion.test_success):
(TestPrintClangVersion.test_failure):
(TestGetLLVMVersion):
(TestGetLLVMVersion.setUp):
(TestGetLLVMVersion.tearDown):
(TestGetLLVMVersion.configureStep):
(TestGetLLVMVersion.test_success):
(TestCheckoutLLVMProject):
(TestCheckoutLLVMProject.configureStep):
(TestCheckoutLLVMProject.test_skipped):
(TestUpdateClang):
(TestUpdateClang.configureStep):
(TestUpdateClang.test_skipped):
(TestUpdateClang.test_use_previous_build):
(TestCheckoutLLVMProject.configureStep.doStepIf): Deleted.
(TestUpdateClang.configureStep.doStepIf): Deleted.

* Tools/CISupport/build-webkit-org/factories.py:
(SaferCPPStaticAnalyzerFactory.__init__):
* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps):

* Tools/CISupport/ews-build/factories.py:
(SaferCPPStaticAnalyzerFactory.__init__): Change the order of steps so we pick
                                          up any changes in the PR.

* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):

Canonical link: <a href="https://commits.webkit.org/299964@main">https://commits.webkit.org/299964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5ccfb67cd82c5b04d3f9020c5f3bc15afaed888

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120690 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40384 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31036 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127083 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72765 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7845db6e-614b-429c-8eab-1ce6b778ec1e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122566 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41081 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48961 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91672 "55 failures") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60923 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/28667ec1-b88d-437f-91f6-7deaa48f4af3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123642 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32805 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108185 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72222 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/473a52c7-77f0-4350-9af1-c41f588393c0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31834 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26292 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70688 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102293 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26474 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129951 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47611 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36149 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100290 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/120110 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47979 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104367 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100129 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45576 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23608 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44271 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19181 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47473 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46942 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50288 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48628 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/Shared/steps_unittest.py:
```diff
@@ -235,21 +260,17 @@ def tearDown(self):
     def test_skipped(self):
         self.configureStep()
-        self.setProperty('llvm_revision', '123456')
+        self.setProperty('curr_llvm_revision', '123456')
```
Nit: `curr_llvm_revision` => `current_llvm_revision`

aj062 commented on Tools/CISupport/Shared/steps.py:
```diff
@@ -190,6 +189,38 @@ def getResultSummary(self):
 
+class GetLLVMVersion(shell.ShellCommandNewStyle, ShellMixin):
+    name = 'get-llvm-version'
+    summary = ''
```
I don't like to set it to empty, as it's not guaranteed that we will update this variable in all scenarios. Even '?' or a placeholder text is fine.

aj062 commented on Tools/CISupport/Shared/steps.py:
```diff
@@ -190,6 +189,38 @@ def getResultSummary(self):
+    summary = ''
+
+    def __init__(self, **kwargs):
+        super().__init__(logEnviron=False, **kwargs)
```
let's add a timeout of (maybe) 1 minute while we are at it.

aj062 commented on Tools/CISupport/Shared/steps.py:
```diff
@@ -190,6 +189,38 @@ def getResultSummary(self):
+
+    @defer.inlineCallbacks
+    def run(self):
+        self.command = self.shell_command('cat Tools/CISupport/safer-cpp-llvm-version')
```
This command will be run on the bot. Is that the behavior you want? Or, you want to read it from the master? (updated when we do `git pull` on master, still wouldn't need buildbot restart).

briannafan replied:
Yes, I want it to be run on the bot! It would be great to test new versions through a PR/EWS and we also would like to be able to update it without having someone on CI intervene.

aj062 commented on Tools/CISupport/Shared/steps.py:
```diff
@@ -190,6 +189,38 @@ def getResultSummary(self):
+
+        rc = yield super().run()
+        if rc != SUCCESS:
+            return defer.returnValue(rc)
```
Do we want to set a fallback for `canonical_llvm_revision` in case this fails? or letting it be None is fine?

briannafan replied:
None is fine. We'll fallback to the previous version if none is found.

aj062 commented on Tools/CISupport/Shared/steps.py:
```diff
@@ -190,6 +189,38 @@ def getResultSummary(self):
+    def getResultSummary(self):
+        if self.results != SUCCESS:
+            self.summary = f'Failed to find canonical LLVM version'
+        return {u'step': self.summary}
```
Let's include the llvm version in the summary in case of success.

briannafan replied:
See line 213! `self.summary = f"Canonical LLVM version: {self.getProperty('canonical_llvm_revision')}"`

aj062 reviewed (approved):
r+ with comments

briannafan committed:
[Safer-CPP] Update LLVM revision without restarting buildbot
https://bugs.webkit.org/show_bug.cgi?id=298691
rdar://160330585

Reviewed by Aakash Jain.

Add GetLLVMVersion to read the canonical LLVM version from a file.
This allows us to update the file and update which version of clang
we're using without restarting buildbot.

* Tools/CISupport/Shared/steps.py:
(GetLLVMVersion):
(GetLLVMVersion.__init__):
(GetLLVMVersion.run):
(GetLLVMVersion.getResultSummary):
(CheckOutLLVMProject.run_vc):
(CheckOutLLVMProject.doStepIf):
(UpdateClang.run):
(UpdateClang.doStepIf):
(PrintClangVersion.run):

* Tools/CISupport/Shared/steps_unittest.py:
(TestPrintClangVersion.test_success):
(TestPrintClangVersion.test_failure):
(TestGetLLVMVersion):
(TestGetLLVMVersion.setUp):
(TestGetLLVMVersion.tearDown):
(TestGetLLVMVersion.configureStep):
(TestGetLLVMVersion.test_success):
(TestCheckoutLLVMProject):
(TestCheckoutLLVMProject.configureStep):
(TestCheckoutLLVMProject.test_skipped):
(TestUpdateClang):
(TestUpdateClang.configureStep):
(TestUpdateClang.test_skipped):
(TestUpdateClang.test_use_previous_build):
(TestCheckoutLLVMProject.configureStep.doStepIf): Deleted.
(TestUpdateClang.configureStep.doStepIf): Deleted.

* Tools/CISupport/build-webkit-org/factories.py:
(SaferCPPStaticAnalyzerFactory.__init__):
* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps):

* Tools/CISupport/ews-build/factories.py:
(SaferCPPStaticAnalyzerFactory.__init__): Change the order of steps so we pick
                                          up any changes in the PR.

* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):

Canonical link: https://commits.webkit.org/299964@main
